src/DFMM.sol

```diff
@@ -278,33 +281,43 @@ contract DFMM is IDFMM {
     // Internals
 
     /**
-     * @dev Transfers `amount` of `token` from the sender to the contract. Note
-     * that if ETH is present in the contract, it will be wrapped to WETH. Any
-     * excess of ETH will be sent back to the sender.
-     * @param token Address of the token to transfer.
-     * @param amount Amount to transfer expressed in WAD.
+     * @dev Transfers `amounts` of `tokens` from the sender to the contract. Note
+     * that if any ETH is present in the contract, it will be wrapped to WETH and
+     * used if sufficient. Any excess of ETH will be sent back to the sender.
+     * @param tokens An array of token addresses to transfer.
+     * @param amounts An array of amounts to transfer expressed in WAD.
      */
-    function _transferFrom(address token, uint256 amount) internal {
-        if (address(this).balance >= amount) {
-            WETH(payable(weth)).deposit{ value: amount }();
+    function _transferFrom(
+        address[] memory tokens,
+        uint256[] memory amounts
+    ) internal {
+        uint256 length = tokens.length;
+
+        for (uint256 i = 0; i < length; i++) {
+            address token = tokens[i];
+            uint256 amount = amounts[i];
 
-            if (address(this).balance > 0) {
-                SafeTransferLib.safeTransferETH(
-                    msg.sender, address(this).balance
-                );
-            }
-        } else {
             uint256 downscaledAmount =
                 downscaleUp(amount, computeScalingFactor(token));
             uint256 preBalance = ERC20(token).balanceOf(address(this));
-            SafeTransferLib.safeTransferFrom(
-                ERC20(token), msg.sender, address(this), downscaledAmount
-            );
+
+            if (token == weth && address(this).balance >= amount) {
+                WETH(payable(weth)).deposit{ value: amount }();
+            } else {
+                SafeTransferLib.safeTransferFrom(
+                    ERC20(token), msg.sender, address(this), downscaledAmount
+                );
+            }
+
             uint256 postBalance = ERC20(token).balanceOf(address(this));
             if (postBalance < preBalance + downscaledAmount) {
                 revert InvalidTransfer();
             }
         }
+
+        if (address(this).balance > 0) {
+            SafeTransferLib.safeTransferETH(msg.sender, address(this).balance);
```

Comment:
Note that this "unconditional ETH return" still potentially represents a DOS vector for interacting contracts that do not have the ability to receive ETH. I don't think it's likely to be a chronic issue as an attacker would have to keep repeatedly sending ETH each time an EOA or ETH-receiving contract interacted w/the DFMM, racking up gas costs. But I could see it being used to block particular contracts from making successful calls in certain time-sensitive situations (e.g. consider a liquidation flow that will sell collateral into a DFMM pool).

The tradeoff of removing this functionality is the inability to recover ETH sent to the contract. I guess that could be a problem if users are trying to trade via raw ETH and aren't sure the exact amount they need to send, so they are generally sending a bit extra which then becomes a loss for them.

So ultimately I would say I agree with keeping this in so long as supporting raw ETH is a design requirement (although it's a requirement I'd question quite strongly).

Reply:
I feel it's best to eliminate this DOS vector if possible. An external integration could be calling `init()`, `allocate()` or `swap()` in a critical function that is expected not to fail.

Assuming that users _can_ calculate the amount of ETH they are supposed to send, you could replace this with a `msg.value == amount` check to prevent users from overpaying.

Alternatively, you could convert the ETH balance into WETH and refund users in WETH instead. Some protocols such as GMX use this approach when they need to refund ETH.

Reply:
In the case of a swap, `amountIn` is specified so it's known and the caller can make `amountIn == msg.value`, but for most allocation functions, there's just a maximum amount. If `DFMM.allocate` enforced a `minDeltaL` then I think the approach of checking `msg.value == amount` would be good, although it limits the design of strategies somewhat.

Refunding in WETH works too, the funds might be unrecoverable if sent to a contract that isn't expecting WETH but at least there's no DOS.

Reply:
We're thinking about following your advice and removing raw ETH completely.

Reply:
This `_transferFrom` call is in a loop though so we'd have to manage the double spending if using msg.value. What if we just remove the DOS vector where we check the balance? What if we only check the contract balance if the token is weth?

Reply:
I guess since we dont have duplicate tokens, we'd only actually hit that branch once in a loop. I'll open a pr

Reply:
PR: https://github.com/primitivefinance/DFMM/pull/122